File: c/2015/02/main.c
#include "../../util.h"

#if 0
static char *FILE_NAME = "sample.txt";
#else
static char *FILE_NAME = "input.txt";
#endif


void part1() {
    ARR_MUL_INTEGERS* in = readIntInputDelim(FILE_NAME, "x");
    int64_t k = 0;
    for (int32_t i = 0; i<in->length; i++) {
        int16_t* s = in->integers[i].integers;
        int16_t a = s[0]*s[1];
        int16_t b = s[1]*s[2];
        int16_t c = s[2]*s[0];
        int16_t w = a<b? (a<c? a : (b<c ? b : c)) : b<c ? b : (a<c ? a : c);
        k += 2*(a + b + c) + w;
    }
    printf("Part 1: %ld\n", k);
    freeArrMulInteger(in);
}

void part2() {
    ARR_MUL_INTEGERS* in = readIntInputDelim(FILE_NAME, "x");
    int64_t k = 0;
    for (int32_t i = 0; i<in->length; i++) {
        int16_t* s = in->integers[i].integers;
        int16_t a = s[0]+s[1];
        int16_t b = s[1]+s[2];
        int16_t c = s[2]+s[0];
        int16_t w = a<b? (a<c? a : (b<c ? b : c)) : b<c ? b : (a<c ? a : c);
        k += s[0]*s[1]*s[2] + 2*w;
    }
    printf("Part 2: %ld\n", k);
    freeArrMulInteger(in);
}

int main(void) {
    part1();
    part2();
    return 0;
}

File: python/2015/02/main.py
#!/usr/bin/env python3

import sys
sys.path.append("../../")
import util as u

###############
debug = False
###############

file = "sample.txt" if debug else "input.txt"

def read_input():
    return u.read_lines(file)

def part1():
    A = read_input()
    k = 0
    for a in A:
        b = [int(x) for x in a.split("x")]
        b.sort()
        k += 3*b[0]*b[1] + 2*(b[1]*b[2] + b[2]*b[0]) 
    print("Part 1:",k)

def part2():
    A = read_input()
    k = 0
    for a in A:
        b = [int(x) for x in a.split("x")]
        b.sort()
        k += 2*b[0] + 2*b[1] + b[0]*b[1]*b[2]
    print("Part 2:",k)

part1()
part2()
